driver/parallel.go

```diff
@@ -0,0 +1,477 @@
+	pg      *parallelGroup
+	scanChs []scanChannel
+	wanted  int
+	dones   IntHeap
+}
+
+func (w *orderedWaiter) Wait() (chan pullResult, error) {
+	for {
+		for _, ps := range w.scanChs {
+			if ps.rank == w.wanted {
+				return ps.ch, nil
+			}
+		}
+		select {
+		case sc, ok := <-w.pg.scanCh:
+			if !ok {
+				return nil, w.pg.scanErr
+			}
+			if sc.ch == nil {
+				w.rankDone(sc.rank)
+				continue
+			}
+			w.scanChs = append(w.scanChs, sc)
+		case <-w.ctx.Done():
+			return nil, w.ctx.Err()
+		}
+	}
+}
+
+func (w *orderedWaiter) rankDone(rank int) {
+	heap.Push(&w.dones, rank)
+	for len(w.dones) > 0 {
+		if w.wanted != w.dones[0] {
+			return
+		}
+		w.wanted++
+		heap.Pop(&w.dones)
+	}
+}
+
+func (w *orderedWaiter) Done(ch chan pullResult) {
+	for i, ps := range w.scanChs {
+		if ps.ch == ch {
+			w.rankDone(ps.rank)
+			w.scanChs[i].ch = nil
+			w.scanChs = append(w.scanChs[:i], w.scanChs[i+1:]...)
+			return
+		}
+	}
+	panic("not reached")
+}
+
+type unorderedWaiter struct {
+	ctx     context.Context
+	pg      *parallelGroup
+	scanChs []scanChannel
+}
+
+func (w *unorderedWaiter) Wait() (chan pullResult, error) {
+	for {
+		if len(w.scanChs) != 0 {
+			return w.scanChs[0].ch, nil
+		}
+		select {
+		case sc, ok := <-w.pg.scanCh:
+			if !ok {
+				return nil, w.pg.scanErr
+			}
+			w.scanChs = append(w.scanChs, sc)
+		case <-w.ctx.Done():
+			return nil, w.ctx.Err()
+		}
+	}
+}
+
+func (w *unorderedWaiter) Done(ch chan pullResult) {
+	for i, ps := range w.scanChs {
+		if ps.ch == ch {
+			w.scanChs[i].ch = nil
+			w.scanChs = append(w.scanChs[:i], w.scanChs[i+1:]...)
+			return
+		}
+	}
+	panic("not reached")
+}
+
+func newParallelScanner(pctx *proc.Context, pg *parallelGroup) *parallelScanner {
+	var w pgWaiter
+	if pg.ordered {
+		w = &orderedWaiter{
+			ctx: pctx,
+			pg:  pg,
+		}
+	} else {
+		w = &unorderedWaiter{
+			ctx: pctx,
+			pg:  pg,
+		}
+	}
+
+	return &parallelScanner{
+		Base: proc.Base{Context: pctx},
+		pg:   pg,
+		w:    w,
+	}
+}
+
+// A parallelScanner is a proc.Proc that generates records with a group
+// of concurrent Scanners, using sources from a MultiSource.
+type parallelScanner struct {
+	proc.Base
+	pg   *parallelGroup
+	w    pgWaiter
+	once sync.Once
+}
+
+func (s *parallelScanner) Pull() (zbuf.Batch, error) {
+	s.once.Do(func() {
+		go s.pg.run()
+	})
+
+	for {
+		ch, err := s.w.Wait()
+		if ch == nil || err != nil {
+			return nil, err
+		}
+
+		select {
+		case pr := <-ch:
+			if proc.EOS(pr.batch, pr.err) {
+				if pr.err != nil {
+					return nil, pr.err
+				}
+				s.w.Done(ch)
+				continue
+			}
+			return pr.batch, nil
+		case <-s.Context.Done():
+			return nil, s.Context.Err()
+		}
+	}
+}
+
+func (s *parallelScanner) Stats() *scanner.ScannerStats {
+	return s.pg.Stats()
+}
+
+type scanSplitResult struct {
+	filterExpr ast.BooleanExpr
+	pre        ast.Proc
+	post       ast.Proc
+}
+
+func scanSplit(program ast.Proc) (scanSplitResult, error) {
+	f, p := liftFilter(program)
+	// todo
```

Comment:
Is my understanding correct that the "todo" here roughly consists of doing what is done by https://github.com/brimsec/zq/commit/7fec0dffdd9c1d224af4f336bf538a7bd2d14490#diff-3b3ea8373a3aa8fc06a84639b6f90779R414 ?

If so, I thought that rather than intermingle this with scanner creation, we would have an AST transform step that, given a parameter N (threads) and information on the source ordering, transforms a linear flowgraph into a parallel one (where possible). This is similar to what we did for spillable groupby. 

(I may not be aware of some constraints and am still digesting this PR..).

Reply:
That's correct about the todo, and how that works & the setup for the parallels lines is something that can change in a subsequent PR. In an earlier revision of this branch, I did create AST nodes for the scanning & merging nodes. (Those have turned into `parallelHead` and `sgWait` in the current revision). That was while I was still working out how to handle the ordering dependencies, and to make it a little easier on myself, I collapsed everything into `createParallelScanner`.

Reply:
This sounds good. I think that I may indeed propose additional changes in a subsequent PR, but given the various threads of work that need to land, I agree that merging this as-is is the practical way to go.